src/team_lib/mhba/xccl_mhba_socket_comm.c

```diff
@@ -0,0 +1,273 @@
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&sockaddr;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, sock_path, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) -1 ] = '\0';
+
+    if (connect(sock, (struct sockaddr *)addr, SUN_LEN(addr)) == -1) {
+        xccl_mhba_error("Failed to connect client socket errno %d", errno);
+        goto fail;
+    }
+    if (do_recvmsg(sock, shared_cmd_fd, shared_pd_handle) != XCCL_OK) {
+        xccl_mhba_error("Failed to recv msg");
+        goto fail;
+    }
+
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close client socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+    return XCCL_OK;
+
+fail:
+    if (close(sock) == -1){
+        xccl_mhba_error("Failed to close client socket errno %d", errno);
+    }
+    return XCCL_ERR_NO_MESSAGE;
+}
+
+static xccl_status_t server_send_data(int command_fd, uint32_t pd_handle, int num_of_connections, xccl_mhba_node_t
+                                        *node, xccl_team_params_t *params, char *sock_path)
+{
+    int sock, i, num_of_curr_connections = 0;
+    struct sockaddr_storage storage= {};
+    struct sockaddr_un *addr;
+
+    connection_t connection[num_of_connections];
+    pthread_t thread[num_of_connections];
+
+    sock = socket(PF_LOCAL, SOCK_STREAM, 0);
+    if (sock == -1){
+        xccl_mhba_error("Failed to create server socket errno %d", errno);
+        return XCCL_ERR_NO_MESSAGE;
+    }
+
+    addr = (struct sockaddr_un *)&storage;
+    addr->sun_family = AF_UNIX;
+    strncpy(addr->sun_path, sock_path, sizeof(addr->sun_path));
+    addr->sun_path[sizeof(addr->sun_path) -1 ] = '\0';
+
+    if (bind(sock, (struct sockaddr *)addr, sizeof(struct sockaddr_un)) == -1) {
+        xccl_mhba_error("Failed to bind server socket errno %d", errno);
+        goto fail;
+    }
+    if (listen(sock, num_of_connections) == -1) {
+        xccl_mhba_error("Failed to listen to server socket errno %d", errno);
+        goto listen_fail;
+    }
+
+    xccl_sbgp_oob_barrier(node->sbgp, params->oob);
+
+    while (num_of_curr_connections < num_of_connections)
+    {
+        /* accept incoming connections */
+        connection[num_of_curr_connections].sock = accept(sock, NULL, 0);
+        connection[num_of_curr_connections].fd = command_fd;
+        connection[num_of_curr_connections].pd_handle = pd_handle;
+        if (connection[num_of_curr_connections].sock != -1)
+        {
+            /* start a new thread but do not wait for it */
+            pthread_create(&thread[num_of_curr_connections], 0, do_sendmsg, (void *)
+            &connection[num_of_curr_connections]);
+            num_of_curr_connections += 1;
+        }
+    }
+
+    for(i=0; i<num_of_connections;i++){
+        pthread_join(thread[i],NULL);
+        if (connection[i].return_val != XCCL_OK){
+            xccl_mhba_error("Failed to send cmd_fd");
+            goto listen_fail;
```

Comment:
i think even in case of listen failure we need to join all other threads, otherwise ASR goes to error flow and destroys the sock file, can this lead to a crash in some of the threads?